[List/ListItem] Introduce selected property #1581

frankf-cgn opened the pull request:
Implementation of Issue #1581:

* following the ReactLink pattern, List has now a selectedLink prop
* selectedLink is a plain js object holding the index of the currently selected ListItem and a callback function to change that value
* ListItem has new prop index
* callback-fn of selectedLink is called when onTouchTap is fired by a ListItem
* the original onTouchTap is propagated, too.

Example implementation for the container component  for the List:

```javascript
export default class MyListContainer extends Component {
  constructor(props) {
    super(props);
    this.state = { selectedIndex: 1 }
  }

  handleUpdateSelectedIndex = (index) => {
    this.setState(Object.assign({}, this.state, {selectedIndex: index}));
  }

  render() {
    var selectedLink = {
      value: this.state.selectedIndex,
      requestChange: this.handleUpdateSelectedIndex
    };
    return(
      <List style={styles.paper} 
        subheader="Contacts"
        selectedLink={selectedLink}>
        <ListItem index={1} ....../>
        <ListItem index={2} ....../>
      </List>
    );
  }
}
```

frankf-cgn commented:
Is there anything I can do to get this merged?

oliviertassinari commented on src/lists/list-item.jsx:
```diff
@@ -48,6 +48,7 @@ const ListItem = React.createClass({
     primaryText: React.PropTypes.node,
     secondaryText: React.PropTypes.node,
     secondaryTextLines: React.PropTypes.oneOf([1, 2]),
+    selected: React.PropTypes.bool,
```
Can you set the default value in getDefaultProps?

oliviertassinari replied:
Shouldn't we add the `index` property too?

oliviertassinari commented on src/lists/list-item.jsx:
```diff
@@ -131,12 +132,13 @@ const ListItem = React.createClass({
     const styles = {
       root: {
-        backgroundColor: (this.state.isKeyboardFocused || this.state.hovered) &&
+        backgroundColor: (this.props.selected && !this.state.hovered) ? selectedColor : ((this.state.isKeyboardFocused || this.state.hovered) &&
```
the `...other` is going to have the `selected` property then passed to the `EnhancedButton`. Is this intentional? Otherwise, can you remove it.

frankf-cgn commented:
I've set the default value as suggested and fixed the unintentionally passing of `selected` to the `EnhancedButton`.

oliviertassinari commented on src/lists/list.jsx:
```diff
@@ -85,16 +88,39 @@ const List = React.createClass({
     }
 
+    let listItems = React.Children.map(children, (child) => {
+      if ( this.props.selectedLink && (child.type.displayName === "ListItem") ) {
```
I would move `this.props.selectedLink` out of the loop.

frankf-cgn replied:
I fully agree.

oliviertassinari commented on src/lists/list.jsx:
```diff
@@ -17,6 +17,9 @@ const List = React.createClass({
 
   propTypes: {
     insetSubheader: React.PropTypes.bool,
+    selectedLink: React.PropTypes.shape({
```
Could you add this to the documentation and with an example?

oliviertassinari commented:
@frankf-cgn Thanks. Could you have a look at my feedback? I think that this PR is going in the right direction :+1:. I will also ask you to squash down the number of commit (to 2 or 3). I will keeep the history clean.

frankf-cgn commented:
@oliviertassinari Thanks for your review and time. I fixed the loop and I'll do the documentation stuff. And of cource I'm going to shrink down the number of commits.

frankf-cgn commented:
@oliviertassinari I have fixed the loop and added the documentation. Concerning the documentation I have tried to choose the wording to be close to the tabs-stuff cos it's somehwat similar to this. Let me know if you are fine with it and I'll clean up the commit history and rebase to the current master.

frankf-cgn commented:
@oliviertassinari Squashed the commits

frankf-cgn commented:
@oliviertassinari If you are ready to merge this, come back to me and I will rebase to the actual master.

oliviertassinari commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -39,6 +39,11 @@ export default class ListsPage extends React.Component {
+  }
+
+  handleUpdateSelectedIndex = (index) => {
+    this.setState(Object.assign({}, this.state, {selectedIndex: index}));
```
Why not doing 
```
this.setState({
  selectedIndex: index,
});
```

frankf-cgn replied:
The way I proposed to set the state ensures that whatever is currently in the state does not get overwritten / thrown away unintentionally, isn't it?

oliviertassinari replied:
As far as I know, we have the same result with the above.

frankf-cgn replied:
Uups, didn't rember that. So I'll change it.

littlefoot32 commented:
Sent from Yahoo Mail on Android

From:"frankf-cgn" <notifications@github.com>
Date:Sat, Nov 7, 2015 at 4:42 AM
Subject:Re: [material-ui] [List/ListItem] Introduce selected property #1581 (#1976)

@oliviertassinari If you are ready to merge this, come back to me and I will rebase to the actual master.

—
Reply to this email directly or view it on GitHub.￼

oliviertassinari commented on src/lists/list-item.jsx:
```diff
@@ -507,6 +511,11 @@ const ListItem = React.createClass({
   },
 
+  _handleTouchTap(e) {
+    this.props.updateSelected(this.props.index);
```
Should we make sur `updateSelected` is defined?

frankf-cgn commented:
@oliviertassinari Thanks for the hints. I have incorporated your suggestions.
One question. In list-item.jsx around line 102, should i add index to const {..., ...other}, so that index is not further propagated to EnhancedButton like you suggested for selected?

oliviertassinari commented:
>  should i add index to const {..., ...other}, so that index is not further propagated to EnhancedButton like you suggested for selected?

Sounds like a good idea

shaurya947 commented:
@oliviertassinari OK to merge?

oliviertassinari commented:
@shaurya947 I didn't try it but it looks good to me.

oliviertassinari commented on src/lists/list.jsx:
```diff
@@ -85,16 +88,45 @@ const List = React.createClass({
     );
   },
+
+  _getSelected(index) {
```
Not sure we need to create two methods since they are called at only on place.

shaurya947 commented:
@oliviertassinari I think you should check it out before you merge since you've reviewed the code (and are more familiar with it).

oliviertassinari commented:
> before you merge

I wasn't going to merge :grin:

shaurya947 commented:
Well, eventually, right? :-P

oliviertassinari commented:
Actually I'm wondering if we shound't create a higer order component to handle selection.
This also somethings needed for the SelectField.

frankf-cgn commented:
@oliviertassinari  I will check out the possibilities to implement this as HOC. Are there any higher order components in the codebase so that I can follow the coding style?

oliviertassinari commented:
@frankf-cgn I'm considering how to implement this. After though, it may not worth it. What do you think?

frankf-cgn commented:
@oliviertassinari Let me investigate on this. You have the arguments on your side: clean separation, reduced complexity and maybe reuse within other components. I'll invest an hour or so into the HOC so that I can tell if it's worth it.

shaurya947 commented:
Thanks @frankf-cgn. Let us know what you find :)

frankf-cgn commented:
Ok, I have something working using a HOC. @oliviertassinari, I think I now get why you hesitated to merge this PR, was it maybe just out of your gut. The point is, that `<List>` in the proposed form is now cluttered with stuff belonging to `<List-Item>` (the whole children-iterator stuff).

## The good
* `<List>` is not needed to be touched
* `<List-Item>`just needs a small patch to have a prop `selected`
* The beef is in the HOC and in the knowledge on how to use it (valueLink)

## The Bad
* HOC needs to have an `onTouchTap`-event/-handler.
* Eventually if this HOC is used with another component (not `<ListItem>`), other handlers might be needed. So in it's current form it is not as reusable as I would like to have it.

## The Code

`src/lists/selectable-enhance.js:`
```
const React = require('react');

export var SelectableEnhance = (Component) => {
  var composed = React.createClass({
    displayName: `Selectable${Component.displayName}`,

    propTypes: {
      index: React.PropTypes.number.isRequired,
      selected: React.PropTypes.bool.isRequired,
      onTouchTap: React.PropTypes.func.isRequired,
    },

    _handleTouchTap(e) {
      this.props.onTouchTap(this.props.index);
    },

    render() {
      return (
        <Component {...this.props} {...this.state } onTouchTap={this._handleTouchTap}/>
      );
    },

  });
  return( composed );
}
```

## The usage example

```
handleUpdateSelectedIndex = (index) => {
    this.setState({
      selectedIndex: index,
    });
}

render() {

  var selectedLink = {
    value: this.state.selectedIndex,
    requestChange: this.handleUpdateSelectedIndex
  };

  return (
    <List
      subheader="Contacts">
    
        <SelectableListItem
          index={1}
          key={1}
    	  selected={this.props.selectedLink.value === 1}
    	  onTouchTap={this.props.selectedLink.requestChange}
          primaryText="Brendan Lim"
          leftAvatar={<Avatar src="images/ok-128.jpg" />} />
        <SelectableListItem 
          index={2}
          key={2}
    	  selected={this.props.selectedLink.value === 2}
    	  onTouchTap={this.props.selectedLink.requestChange}
          primaryText="Grace Ng"
          leftAvatar={<Avatar src="images/uxceo-128.jpg" />} />
        <SelectableListItem 
          index={3}
          key={3}
    	  selected={this.props.selectedLink.value === 3}
    	  onTouchTap={this.props.selectedLink.requestChange}
          primaryText="Kerem Suer"
          leftAvatar={<Avatar src="images/kerem-128.jpg" />} />
    
    </List>
  );
}
```

If we schould follow this route I'm going to update my PR.

oliviertassinari commented:
@frankf-cgn Thanks for exploring this path!
Looking at your example, I can't see the problem `selectable-enhance.jsx` is adressing.

frankf-cgn commented:
Okay, I have to think about this. If this is the case, the only thing that is really needed is the selected-prop within `<ListItem>`.

oliviertassinari commented on src/lists/list.jsx:
```diff
@@ -85,16 +88,45 @@ const List = React.createClass({
+    return this.props.selectedLink.value === index;
+  },
+
+  _updateSelectedIndex(index) {
```
Actually, I would keep this method to keep the same reference for the `updateSelected` property.
This can save use some rerendering when using the purerender mixin.

shaurya947 commented:
I think wrapping *every* `ListItem` in a HOC is not the best idea as multiple props are repeated across all the `ListItem`s (`onTouchTap`, `selected`, `index` and `key` that could be generated sequentially).

@frankf-cgn what do you think about a HOC that wraps the whole `List`? How re-usable would that be?

If this seems like a shot in the dark, then I agree with @oliviertassinari: the original state of the PR is what we should do.

frankf-cgn commented:
@shaurya947 are you on gitter or discord?

frankf-cgn commented:
ok, I have a working List-HOC. Without changing `<List>` and `<LisItem>` at all and heavy "borrowing" from `material-ui/src/menus/menus/menu.jsx` the following code emerged.
Despite my concerns that this HOC might not be reusable, i think it is. In fact, I think that with some more work, this HOC can used in `material-ui/src/menus/menus/menu.jsx`, too.

## The HOC

```javascript
const React = require('react');
const StylePropable = require('material-ui/lib/mixins/style-propable');

export var SelectableContainerEnhance = (Component) => {
  var composed = React.createClass({
    mixins: [StylePropable,],
    displayName: `Selectable${Component.displayName}`,

    propTypes: {
      valueLink: React.PropTypes.shape({
        value: React.PropTypes.number,
        requestChange: React.PropTypes.func
      }).isRequired,
    },

    getValueLink: function(props) {
      return props.valueLink || {
        value: props.value,
        requestChange: props.onChange
      };
    },

    render(){
      var { children } = this.props
      let listItems;
      let keyIndex = 0;

      const styles = {
        selectedItem: {
          backgroundColor: '#3fa43a'
        }
      };

      listItems = React.Children.map(children, (child) => {
        if (child.type.displayName === "ListItem") {
          let selected = this._isChildSelected(child, this.props);
          let selectedChildrenStyles = {};
          if (selected) {
            selectedChildrenStyles = styles.selectedItem;
          }

          let mergedChildrenStyles = this.mergeStyles(
            child.props.style || {},
            selectedChildrenStyles
          );

          keyIndex += 1;

          return React.cloneElement(child, {
            onTouchTap: (e) => {
              this._handleItemTouchTap(e, child);
              if (child.props.onTouchTap) { child.props.onTouchTap(e) };
            },
            key: keyIndex,
            style: mergedChildrenStyles,
          });
        }
        else {
          return child;
        }
      });
      children = listItems;

      return (
        <Component {...this.props} {...this.state} children={children}  />
      );
    },

    _isChildSelected(child, props) {
      let itemValue = this.getValueLink(props).value;
      let childValue = child.props.value;

      return (itemValue && itemValue === childValue);
    },

    _handleItemTouchTap(e, item) {
      let valueLink = this.getValueLink(this.props);
      let itemValue = item.props.value;
      let menuValue = valueLink.value
      if ( itemValue !== menuValue) { 
        valueLink.requestChange(e, itemValue);
      }
    },

  });
  return( composed );
};

```

## The usage example

```javascript
handleUpdateSelectedIndex = (e,index) => {
    this.setState({
      selectedIndex: index,
    });
}

render() {
var SelectableList = SelectableContainerEnhance(List);

  var valueLink = {
    value: this.state.selectedIndex,
    requestChange: this.handleUpdateSelectedIndex
  };

  return (
    <SelectableList
      valueLink={valueLink}
      subheader="Contacts">
    
        <ListItem
          value={1}
          primaryText="Brendan Lim"
          leftAvatar={<Avatar src="images/ok-128.jpg" />} />
        <ListItem 
          value={2}
          primaryText="Grace Ng"
          leftAvatar={<Avatar src="images/uxceo-128.jpg" />} />
        <ListItem 
          value={3}
          primaryText="Kerem Suer"
          leftAvatar={<Avatar src="images/kerem-128.jpg" />} />
    
    </SelectableList>
  );
}
```

oliviertassinari commented:
@frankf-cgn Yeah, I think that it's much better like this!

frankf-cgn commented:
I really like it this way, too! 
How should we move on from this point? If you like, I'm going to update my branch `issue/1581` with this HOC and reverting back the changes on `<List>` an `<ListItem>`. Two question remain then:
1. Where to put HOCs in the file structure of material-ui
2. How to Name this HOC

shaurya947 commented:
```js
getValueLink: function(props) {
      return props.valueLink || {
        value: props.value,
        requestChange: props.onChange
      };
    },
```

^like the flexibility here.

```js
return React.cloneElement(child, {
            onTouchTap: (e) => {
              this._handleItemTouchTap(e, child);
              if (child.props.onTouchTap) { child.props.onTouchTap(e) };
            },
            key: keyIndex,
            style: mergedChildrenStyles,
          });
```

^so the developer can override the `onTouchTap` for specific ListItems too? That's cool.


---------------

@frankf-cgn this is some pretty good stuff. You implemented the idea well. I would say yes, go ahead and update your PR to use this code **and** update the docs if you can. To answer your questions:

> 1.Where to put HOCs in the file structure of material-ui

Since we're aiming to make this component more general / re-usable, I would say just create a folder `hoc` under `src` and put it there.

> 2.How to Name this HOC

What you're calling it right now is fine (`SelectableContainerEnhance`).

As for the docs, I can see at least two approaches:

- Create a new section called "HOC" under "Components" OR
- Add documentation for this feature to the "Lists" page.

The first approach might be better if we're planning on making this more general than lists (and perhaps adding more HOCs in the future). What are your thoughts on 1., 2. and the docs? Also, @oliviertassinari

frankf-cgn commented:
@shaurya947  As I said before, I heavily "borrowed" from `material-ui/src/menus/menus/menu.jsx` and as I don't want to adorn myself with false laurels, praise and thank belong to the author of `material-ui/src/menus/menus/menu.jsx`! 
I have ssen the valueLink stuff first here https://gist.github.com/BinaryMuse/bd66e40b00598e1288f2#file-ex_09-js, and the stuff in `material-ui/src/menus/menus/menu.jsx` seems to be inspired from this, but who knows, mybe it's from the same author ;-)

Nevertheless, I will create the folder `src/hoc ` and put the code in there. 
As for the docs I think it makes a lot of sense to initialy put this under the List page since the first usage of the hoc is there. Later on adding a section components/HOC might be a good idea.

shaurya947 commented:
Fine by me :+1:

frankf-cgn commented:
I have just updated the PR with the discussed changes and some minor improvements:
* `selectedItemStyle` prop with fallback to...
* ThemeSupport based default

#### Performance of `<ListsPage>`
Because the performance in dev-mode of the doc page was not as good as it could be, I have introduces a further hoc with `docs/src/app/components/pages/components/lists.jsx` holding the state for the `<SelectableList>`. I don't think that this hoc should be introduced generally, b/c if you use  `<SelectableList>` in the context of flux/redux or if you have only a few components in the parent everthing is fine perf-wise.

#### Documentation strategy
I have had hard time to document this stuff b/c I think some concepts have to be explained in more detail. Is there any component or otherwise 'default' way to do? Since I have not found sth, I have introduced a markdownElement, but this is not the best option either, especially not in the context of this issue. Hate to have unrelated stuff in a PR.

shaurya947 commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -54,6 +85,12 @@ export default class ListsPage extends React.Component {
           {
+            name: 'selectedItemStyle',
+            type: 'object',
+            header: 'optional, only available it HOC SelectableContainerEnhance is used',
```
> only available **if** HOC

shaurya947 commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -71,6 +108,13 @@ export default class ListsPage extends React.Component {
+          {
+            name: 'valueLink',
+            type: 'valueLink',
+            header: 'optional, only available it HOC SelectableContainerEnhance is used',
```
same here

shaurya947 commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -34,11 +36,40 @@ const { Colors } = Styles;
 const CodeBlock = require('../../code-example/code-block');
+let SelectableList = SelectableContainerEnhance(List);
+
+function wrapState(ComposedComponent) {
```
Is there any way we could avoid having this helper function to wrap `SelectableList`? And do it straight using props in the render method of `ListsPage`? I think it would be a good idea to keep the docs code as straightforward as possible..

frankf-cgn replied:
Yeah, it is avoidable, but it makes the `<ListsPage>` really slow. There is a noticable lag if you click on an item of the `<SelectableList>`. See the performance-Section of my previous comment.

shaurya947 commented on docs/src/app/components/pages/home.jsx:
```diff
@@ -8,6 +8,7 @@ const { StylePropable, StyleResizable } = Mixins;
 const { Colors, Spacing, Typography } = Styles;
 const ThemeManager = Styles.ThemeManager;
 const DefaultRawTheme = Styles.LightRawTheme;
+require('github-markdown-css/github-markdown.css');
```
This statement throws an error on my side. Not sure if you can import CSS like that. Might have to include in the `index.html`

frankf-cgn replied:
You can do it this way if you have a webpack loader for style configured, like I did it in `docs/webpack-dev-server.config.js`and `docs/webpack-production-server.config.js`. Might be that you must `npm install` first.
Anyway this is purely academic, because I'm going to remove this stuff as I want to stick with the existing documentation standard. (you're last comment....)

shaurya947 commented:
@frankf-cgn thus far we've used simple html tags with some `CodeBlock` components for documentation. I think we should try to stick to that for consistency and avoid the whole markdown thing. See [here](https://github.com/callemall/material-ui/blob/master/docs/src/app/components/pages/get-started/installation.jsx) for an example.

frankf-cgn commented:
The typos are fixed, some rewording has taken place and I have removed the markdown stuff. Hope we have reached the last itterations ;-)

shaurya947 commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -666,9 +742,105 @@ export default class ListsPage extends React.Component {
+          <ul>
+            <li> enhance <code>&lt;List&gt;</code> with HOC</li>
+            <li> decide where to put state</li>
+            <li> implement end set valueLink</li>
```
should this read "implement *and* set valueLink" ?

shaurya947 commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -666,9 +742,105 @@ export default class ListsPage extends React.Component {
+          </p>
+          <p>
+            A possible solution for this is to use another hoc. An example can be found in the sourcecode of 
+            <code>docs/src/app/components/pages/components/lists.jsx</code>.
```
docs don't show a space right before the `<code>` tag begins..

shaurya947 commented on src/hoc/selectable-enhance.js:
```diff
@@ -0,0 +1,123 @@
+      let newChildren = listItems;
+
+      return (
+        <Component {...this.props} {...this.state} children={newChildren}  />
```
I feel like we should change this to:

```js
<Component {...this.props} {...this.state}>
    {newChildren}
</Component>
```

frankf-cgn replied:
Wait, there is something fishy going in here - my code looks completley different from whats in here. Let me investigate on this.

frankf-cgn replied:
That looks al ot better!

shaurya947 commented:
I just pointed out a few quick things @frankf-cgn. Once you can take care of them, we should merge this.

I'll also give @oliviertassinari a chance to look over this.

frankf-cgn commented:
@shaurya947 I was so sure that I have find all typos during my proofreading. Ok, everthing done by now and pushed.

oliviertassinari commented:
@shaurya947 Let's see

oliviertassinari commented on docs/src/app/components/pages/components/lists.jsx:
```diff
@@ -666,9 +742,105 @@ export default class ListsPage extends React.Component {
+          </p>
+          <div style={styles.codeblock}>
+            <CodeBlock>
+{`import { SelectableContainerEnhance } from 'material-ui/hoc/selectable-enhance';
```
missing `lib`  ?

oliviertassinari commented on src/hoc/selectable-enhance.js:
```diff
@@ -0,0 +1,124 @@
+              this._handleItemTouchTap(e, child);
+              if (child.props.onTouchTap) { child.props.onTouchTap(e) };
+            },
+            key: keyIndex,
```
> key and ref from the original element will be preserved

From the React doc. Shouldn't we remove this?

frankf-cgn replied:
Is there any benefit for the user of the component to let him set the key which I do not see? Otherwise I would like to mitigate the "boilerplate" and source of error/warnings here.

oliviertassinari replied:
You are right.
Second question then. Do we need a key at all here? For an array we do, but since you are using `React.Children.map` I'm not sure.

frankf-cgn replied:
@oliviertassinari Quite frankly, I didn't know. `React.Children.map` does iterate over children and returns an array, thats for sure. 
But adding the key anyway has the benefit, that for the usecase "`<ListItems>` generated by loop" the user has not to remember adding the key.
Somehow in the codebase of `<ListsPage>`this problem already exist. If you run the doc in development env, you get this nasty `Warning: Each child in an array or iterator should have a unique "key" prop`. (I couldn't track down the reason of this, but if I find,  I will create an issue and submit a PR :-) )

oliviertassinari replied:
@frankf-cgn Let's keep the key then. Can you have a look at my comment on the import? I think that we can merge then :+1:.

oliviertassinari commented:
@frankf-cgn Thank you!

frankf-cgn commented:
@oliviertassinari @shaurya947 I have to thank you both!

shaurya947 commented:
Yay! Great job @frankf-cgn
